src/zalsa.rs

```diff
@@ -130,6 +132,9 @@ pub struct Zalsa {
     /// adding new kinds of ingredients.
     jar_map: Mutex<FxHashMap<TypeId, IngredientIndex>>,
 
+    /// Map from the type-id of a salsa struct to the index of its first ingredient.
+    salsa_struct_map: Mutex<FxHashMap<TypeId, IngredientIndex>>,
+
```

Comment:
I don't like this new table for `SalsaStruct` -> `IngredientIndex` but I couldn't find any other way to pass the salsa-struct-ingredient to its tracked function when creating the function's ingredient, as salsa-struct-ingredient is accessible only with its `Configuration`, that is private inside the macro-expanded code of that salsa struct.
Would there be some good alternatives?

Reply:
I would be inclined to add a method `ingredient_index` to `SalsaStructInDb` but best to hear what @nikomatsakis thinks

Reply:
We need to get the `IngredientIndex` for salsa struct here;

https://github.com/salsa-rs/salsa/blob/e4d36daf2dc4a096009753b6e326cf44e6c6e18a/components/salsa-macro-rules/src/setup_tracked_fn.rs#L196-L205

and we can't get the actual instance of salsa struct here. All we have is `JarAux` - actually, this is `Zalsa` -, so that function should be a static method, I guess. (I'm afraid that this makes `SalsaStructInDb` dyn-incompatible but we are not using it as a trait object anywhere in salsa)

Thus, the function would be implemented inside our `setup_*` macros like;

```rust
fn ingredient_index(aux: &dyn JarAux) -> IngredientIndex {
    aux.lookup_jar_by_type(JarImpl::<$Configuration>::default())
}
```

and we need to expose some method like `lookup_jar_by_type` in `JarAux`, I think.